### ISSUE TITLE
Feature Request: source generator

### ISSUE THREAD

**Issue by jetersen** [updatecli/updatecli#774]
### Is your feature request related to a problem?

So we have Jenkins Plugin Manager and instead of porting it to golang and also another problem having to declaratively specify everything in updatecli.

I feel like we could use a source generator pattern. What if we could from a pom.xml generate a list of sources by checking each dependency? Much like dependabot without having to declare each, because a pom.xml is already declarative and with plus 10-30 maven dependency updatecli config can become verbose.

Would be nice to have the source generator and than the ability to check via shell whether it supports Jenkins Version x

### Solution you'd like

generator pattern for sources!

### Alternatives you've considered

Currently what we are doing in https://github.com/jenkinsci/bom/blob/master/updatecli/updatecli.d/update-plugins.yaml

**Comment by olblak** [comment]
I am currently working on a new Updatecli feature, named "autodiscovery". The goal of the autodiscovery is to generate manifest based on a context. So by default running `updatecli apply/diff/show` would generate the manifest and then run them right away.

The GitHub issue to track this work is https://github.com/updatecli/updatecli/issues/761
I already have a working pullrequest but I am missing the documentation and tests, and probably some edge case that I need to validate
As a first iteration I am automating Helm chart dependency updates but I could easily add more later.

As you mention, when we know what to look for it's easy to generate the manifest.

**Comment by jetersen** [comment]
Okay autodiscovery would be awesome. :)

**Comment by jetersen** [comment]
Basically the issue that I have is I want my pull request from update cli to be properly titled and potentially also have one PR for each update of jenkins plugin.

At the moment we have things like this: `Bump pinned plugins` https://github.com/jenkinsci/bom/pull/1308
I could improve the title using multiple PRs for each line but I am lacking the ability to understand that each plugin should be a separate bump within that line.

**Comment by jetersen** [comment]
Of course I could generate the manifest myself with some powershell, now that I think about it as interim solution.

**Comment by dduportal** [comment]
> I don't know if it makes sense to port plugin-manager to golang. I know during talks about plugin-manager creation there were talks about using golang at the time.

That was the goal of https://github.com/jenkins-infra/uc but it has been archived in favor of the `jenkins-plugin-cli`.
The reason is that the cost of rewriting another dependency resolver in another language has a really high maintenance cost, because it has to keep up with both Jenkins Core and the `jenkins-plugin-cli` which are already subtely different.

The problem you're raising @jetersen is absolutely legit (we... have the same on the Jenins infra). 

For context on the Jenkins infra, we decided to stop caring about updating plugins atomically and switch to a "update often so each update is a small increment easier to rollback and diagnose if it fails". One of the arguments behind this is that almost no Jenkins plugin follows a proper semver, it's more a "keep up to latest". 
So we resort to a shell script: https://github.com/jenkins-infra/docker-jenkins-lts/blob/main/bin/update-plugins.sh that we plan to move to an updatecli shell target.

But that is also the goal of resource plugins in updatecli: if you are feeling up to the challenge, writing (and maintaining) plugins for each of the Jenkins Plugins / Maven / etc. domains could be useful to a lot of people.

**Comment by jetersen** [comment]
Regarding a resource plugin for updatecli I am afraid of the problems complexity.

Not all plugins have the same release process so it makes it hard to do it right.
Like changelog is different for many plugins, of course we could favor GitHub releases.

Integrating with the jenkins update center is the easy part.
However then you have potential excluded plugins in the update center.
Then there is backporting which is not well established either.

I think once autodiscovery lands it might be easy to have a pom with dependencies or a plugin.txt be parsed.
Having to manually write updatecli would be annoying for that many plugins.

Potentially we already have maven repository but than we could use a condition for jenkins update center but could also be source.

**Comment by jetersen** [comment]
I know the time would properly be spent in golang.

But I got there for bom with powershell:
https://github.com/jenkinsci/bom/runs/7482133253?check_suite_focus=true#step:6:968

**Comment by jetersen** [comment]
https://github.com/jenkinsci/bom/pull/1314 ya first attempt at generation got updatecli confused 😆

**Comment by olblak** [comment]
Now that the xml resource is released ([0.32.0](https://github.com/updatecli/updatecli/releases/tag/v0.32.0)),  ready and working quite well on my project. I started thinking to the current issue but I don't think I am understanding it well.

While parsing all pom.xml, I could easily handle dependencies listed in `//project/dependencies/` based on repositories configured in `//project/repositories` ( + maven central as fallback)

Would it makes sense to also doing it for DependencyManagement like `//Project/DependencyManagement/Dependencies`

**Comment by jetersen** [comment]
Yes absolutely also for dependency management

**Comment by olblak** [comment]
I opened a draft pullrequest here https://github.com/updatecli/updatecli/pull/833

Right now it only support "dependencies"

> Does it support reading a parent pom ?
Do you mean updating something like this https://github.com/jenkins-infra/pipeline-steps-doc-generator/blob/master/pom.xml#L4L8 ? 
If yes then I could add it as well.

I decided to ignore properties such as <version>${myProperty}</version>

I don't know the difference between dependencies and dependencyManagement, but I guess for each of them, I could also retrieve the latest version from Maven repository

**Comment by olblak** [comment]
The way the autodiscovery works is either you provide no configuration such 

`./bin/updatecli diff --experimental `: In this case it search locally and recursively available updates but no scm support
`./bin/updatecli diff --config e2e/updatecli.d/warning.d/autodiscovery/maven/git.yaml --experimental `: On top of adding additional parameter, it allows to clone a git repository in /tmp/ and then search there available updates and commit/push/pullrequest if needed

**Comment by olblak** [comment]
The pullrquest is now ready and support the following cases


1. Read pom.xml looking and try to bump version in `/project/parent` if a version is specified and do not contains a property
2. Read pom.xml looking and try to bump version in `/project/dependencies/*` if a version is specified and do not contains a property 
3. Read pom.xml looking and try to bump version in `/project/dependencyManagement/dependencies/*` if a version is specified and do not contains a property

**Comment by olblak** [comment]
I think this issue has been solved with the maven autodiscovery feature
-> https://www.updatecli.io/docs/plugins/autodiscovery/maven/

I know documentation is lacking so feel free to reopen this issue if you need more information.
